tv_apps/tv_shared/js/animations.js

```diff
@@ -72,6 +72,93 @@
         }
       };
+    },
+
+    _getChild: function(container, childRelation, index) {
+      var targetElem = container.children[index];
+      childRelation = childRelation.child;
+      while(childRelation.child) {
+        targetElem = targetElem.children[0];
+        childRelation = childRelation.child;
+      }
+      return targetElem;
+    },
+
+    _findChildInViewport: function(container, childRelation) {
+      // find the elements in viewport
+      var child;
+      var childRect;
+      var childrenInViewport = [];
+      var i;
+      for(i = 0; i < container.children.length; i++) {
+        child = this._getChild(container, childRelation, i);
+        childRect = child.getBoundingClientRect();
+        if (childRect.right >= 0 &&
+          childRect.left <= window.innerWidth) {
+          childrenInViewport.push({
+            left: childRect.left,
+            child: child
+          });
+        } else {
+          child.classList.remove('animation-bubble-init');
+        }
+      }
+
+      // sort child indexes. Card position may not be the same as index
+      // after a new child is pinned to home
+      childrenInViewport.sort(function(childObj1, childObj2) {
+        if (childObj1.left > childObj2.left) {
+          return 1;
+        }
+        if (childObj1.left < childObj2.left) {
+          return -1;
+        }
+        return 0;
+      });
+
+      return childrenInViewport;
+    },
+
+    doBubbleAnimation: function(container, childRelation, bubbleInterval,
+                                callback) {
+
+      var childrenInViewport = this._findChildInViewport(
+                                            container, childRelation);
+
+      // start bubbling animation of every child
+      var childTransitionEndCount = 0;
+      var child;
+      var i;
+      for(i = 0; i < childrenInViewport.length; i++) {
+        child = childrenInViewport[i].child;
+        child.classList.add('animation-bubble-start');
+        child.classList.add('animation-bubble-paused');
+        // jshint loopfunc:true
+        (function(child) {
+          setTimeout(function() {
+            // bubbling transition
+            child.classList.remove('animation-bubble-paused');
+            var bubbleEnd = function(evt) {
+              if (evt.target === child) {
+                childTransitionEndCount++;
+                child.removeEventListener('animationend', bubbleEnd);
+
+                // last element has finished bubbling
+                if (childTransitionEndCount === childrenInViewport.length) {
+                  for(i = 0; i < childrenInViewport.length; i++) {
+                    child = childrenInViewport[i].child;
+                    child.classList.remove('animation-bubble-start');
+                  }
+                  if (callback) {
+                    callback();
+                  }
+                }
+              }
+            };
+            child.addEventListener('animationend', bubbleEnd);
```

Comment:
Can we just make use of `style.animationDelay` of each children rather than nesting `setTimeout`?